### PR TITLE
Allows bulk instance state changes.

### LINKED ISSUE #7934: Bulk instance state change API
There has been interest in making `lxc start --all` and similar stop/restart/pause commands be handled server-side rather than client side.

This would also allow those to respect the `boot.autostart.priority`, boot.autostart.delay` and `boot.stop.priority` properties either out of the box or when provided an extra flag.

The current plan is to implement this through `PUT /1.0/instances` taking a struct similar to that we use for `PUT /1.0/instances/NAME/state`.

### PR DESCRIPTION
**Pull request by kevtheappdev**
Fixes #7934

### CODE REVIEW

File: doc/api-extensions.md [hunk 1]
```diff
@@ -1257,3 +1257,8 @@ with `ovn.ingress_mode=routed`.
 ## projects\_limits\_instances
 Adds `limits.instances` to the available project configuration keys. If set, it
 limits the total number of instances (VMs and containers) that can be used in the project.
+
+## instances
```
Comment (maxpat21) [inline doc/api-extensions.md:1261]: We were unsure whether or not we should include an entry here, as well as what the header for it should be, should we also include a new entry in shared/version/api.go?

File: lxd/instances_put.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+	var opType db.OperationType
+	var do func(*operations.Operation) error
+	switch action {
+	case shared.Start:
+		opType = db.OperationContainerStart
+		do = func(op *operations.Operation) error {
+			for _, inst := range instances {
+				inst.SetOperation(op)
+				if err = inst.Start(raw.Stateful); err != nil {
+					return err
```
Comment (kevtheappdev) [inline lxd/instances_put.go:68]: How should we convey error state here? If one instance fails to set state we will exit and not set state for any of the others after it. I wasn't sure how to go about "collecting" errors for all these bulk state changes, or how to immediately convey one without breaking iteration.

File: lxc/action.go [hunk 3]
```diff
@@ -231,30 +290,13 @@ func (c *cmdAction) Run(cmd *cobra.Command, args []string) error {
...
-					}
-				}
-				names = append(names, fmt.Sprintf("%s:%s", resource.remote, ct.Name))
-			}
+		err = c.doActionAll(cmd.Name(), resources)
+		if err != nil {
+			return err
 		}
 	} else {
 		names = args
```
Comment (kevtheappdev) [inline lxc/action.go:302]: If the user passes in multiple instance names (I.e. `lxc start thing1 thing2 thing3`), should this processing also happen on the backend? We designed this new endpoint focusing on the case when the user explicitly gives `--all`

File: doc/api-extensions.md (hunk 1 above)
Comment (tomponline) [reply doc/api-extensions.md:1261]: Yes please do include the entry in shared/version/api (as part of the same commit that adds the modification to this file).

Also the API extension name `instances` is too vague as it does not explain what is being added, suggest something like `instance_bulk_state_change`.

File: lxd/instances_put.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+	"github.com/lxc/lxd/lxd/operations"
+	"github.com/lxc/lxd/lxd/response"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/api"
+	"github.com/lxc/lxd/lxd/cgroup"
+)
+
+
+/*
+* Update state in bulk
```
Comment (tomponline) [inline lxd/instances_put.go:20]: We normally just use `// comments`

File: lxd/instances_put.go (hunk 2 above)
Comment (tomponline) [reply lxd/instances_put.go:68]: Its a good question, you can certainly collect a slice of errors and then process them at the end (perhaps concatenating them into a single multi-line error).

@stgraber is there any convention around handling "partial" error responses from the API?

File: client/interfaces.go [hunk 5]
```diff
@@ -87,6 +87,7 @@ type InstanceServer interface {
 	// Container functions
 	GetContainerNames() (names []string, err error)
 	GetContainers() (containers []api.Container, err error)
+	PutInstances(state api.InstancesPut, ETag string) (Operation, error)
```
Comment (stgraber) [inline client/interfaces.go:90]: Should be `UpdateInstances` for consistency.

File: client/interfaces.go [hunk 6]
```diff
@@ -87,6 +87,7 @@ type InstanceServer interface {
 	// Container functions
 	GetContainerNames() (names []string, err error)
 	GetContainers() (containers []api.Container, err error)
+	UpdateInstances(state api.InstancesPut, ETag string) (Operation, error)
```
Comment (stgraber) [inline client/interfaces.go:90]: return should be written as `(op Operation, err error)` for consistency with the rest.

File: doc/rest-api.md [hunk 7]
```diff
@@ -1434,7 +1451,7 @@ Input:
     "action": "stop",       // State change action (stop, start, restart, freeze or unfreeze)
     "timeout": 30,          // A timeout after which the state change is considered as failed
     "force": true,          // Force the state change (currently only valid for stop and restart where it means killing the instance)
-    "stateful": true        // Whether to store or restore runtime state before stopping or startiong (only valid for stop and start, defaults to false)
+    "stateful": true        // Whether to store or restore runtime state before stopping or starting (only valid for stop and start, defaults to false)
```
Comment (stgraber) [inline doc/rest-api.md:1454]: Can you split this one into its own commit? That will let us easily cherry-pick it to our stable branches.

File: shared/version/api.go [hunk 8]
```diff
@@ -244,6 +244,7 @@ var APIExtensions = []string{
 	"network_physical_routes_anycast",
 	"projects_limits_instances",
 	"network_state_vlan",
+	"instance_bulk_state_change",
```
Comment (stgraber) [inline shared/version/api.go:247]: This needs to go into the same commit as the `doc/api-extensions.md` change.

File: shared/api/instance.go [hunk 9]
```diff
@@ -28,6 +28,13 @@ type InstancesPost struct {
 	Type         InstanceType   `json:"type" yaml:"type"`
 }
 
+type InstancesPut struct {
```
Comment (stgraber) [inline shared/api/instance.go:31]: This needs a docstring comment above it or `golint` will fail.

Also, this needs another comment to indicate the API extension introducing it, look at `InstancesPost` below for an example.

File: lxd/instances_put.go [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+	resources := map[string][]string{}
+	resources["containers"] = names
+
+	op, err := operations.OperationCreate(d.State(), project, operations.OperationClassTask, opType, resources, nil, do, nil, nil)
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	return operations.OperationResponse(op)
+}
```
Comment (stgraber) [inline lxd/instances_put.go:233]: Missing final line break somehow?

File: lxd/instances.go [hunk 11]
```diff
@@ -27,6 +27,7 @@ var instancesCmd = APIEndpoint{
 
 	Get:  APIEndpointAction{Handler: containersGet, AccessHandler: allowProjectPermission("containers", "view")},
 	Post: APIEndpointAction{Handler: containersPost, AccessHandler: allowProjectPermission("containers", "manage-containers")},
+	Put:  APIEndpointAction{Handler: containersPut, AccessHandler: allowProjectPermission("containers", "manage-containers")},
```
Comment (stgraber) [inline lxd/instances.go:30]: This should be `operate-containers` instead for this particular endpoint.

File: lxd/instances_put.go [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+			}
+
+			return coalesceErrors(failures)
+		}
+	default:
+		return response.BadRequest(fmt.Errorf("unknown action %s", raw.Action))
+	}
+
+	resources := map[string][]string{}
+	resources["containers"] = names
```
Comment (stgraber) [inline lxd/instances_put.go:225]: Let's make that `"instances"` rather than `"containers"`.

File: lxd/instances_put.go [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+				res := inst.Unfreeze()
+				if res != nil {
+					failures[inst.Name()] = res
+				}
+			}
+
+			return coalesceErrors(failures)
+		}
+	default:
+		return response.BadRequest(fmt.Errorf("unknown action %s", raw.Action))
```
Comment (stgraber) [inline lxd/instances_put.go:221]: We usually capitalize our error strings, so should be `"Unknown action %s"`

File: lxc/action.go (hunk 3 above)
Comment (stgraber) [reply lxc/action.go:302]: Nope, for now we'll just care about `--all`.

File: lxc/action.go [hunk 14]
```diff
@@ -231,30 +290,13 @@ func (c *cmdAction) Run(cmd *cobra.Command, args []string) error {
 			return err
 		}
 
-		for _, resource := range resources {
```
Comment (stgraber) [inline lxc/action.go:234]: So one thing to keep in mind too is that the CLI needs to be backward compatible.

So you'll instead want to keep the code as it is and add a short path like:
```
if c.flagAll && d.HasExtension("name_of_extension") {
    return doActionAll(...)
}
```

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
@tomponline can you review this tomorrow and merge if you're happy with it?

I've fixed all the review comments we had on this, fixed a few unrelated things here and there, added parallel run on the server side (to match the client logic we had) and tweaked error handling a bit too.

File: lxd/instances_put.go [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+				continue
+			}
+		}
+
+		instances = append(instances, inst)
+		names = append(names, inst.Name())
+	}
+
+	// Don't mess with containers while in setup mode
+	<-d.readyChan
```
Comment (tomponline) [inline lxd/instances_put.go:80]: @stgraber what is "setup" mode in this context (for my own info)?

File: lxd/instances_put.go [hunk 16]
```diff
@@ -0,0 +1,319 @@
...
+			failures := map[string]error{}
+			failuresLock := sync.Mutex{}
+			wgAction := sync.WaitGroup{}
+
+			for _, inst := range instances {
+				wgAction.Add(1)
+				go func(inst instance.Instance) {
+					defer wgAction.Done()
+
+					ephemeral := inst.IsEphemeral()
```
Comment (tomponline) [inline lxd/instances_put.go:204]: @stgraber not strictly related to the PR, but it seems like this logic should be moved into the Instance `Restart()` functions (or a common function called from the per-driver Restart() functions), as this feels like internal logic that the instance should be taking care of, rather than the 'user' of the Restart() function having to know about (I write this as someone who wouldn't have thought to handle the ephemeral case for Restart). What do you think?

File: lxd/instances_put.go [hunk 17]
```diff
@@ -0,0 +1,319 @@
...
+			}
+
+			wgAction.Wait()
+			return coalesceErrors(failures)
+		}
+	case shared.Stop:
+		opType = db.OperationInstanceStop
+		if raw.Stateful {
+			do = func(op *operations.Operation) error {
+				failures := map[string]error{}
```
Comment (tomponline) [inline lxd/instances_put.go:114]: @stgraber an observation here is that there's quite a lot of repetition on the `failures`, `failuresLock` and `wgAction` setup and usage. It may have been easier to initialise a buffered channel using a temporary struct with name and error result and then posted the result of each instance's action into it and then collect them at the end. But I note there do seem to be a few special cases further down that may have made this approach unfeasible.

**Review by tomponline — APPROVED** [review]

File: lxd/instances_put.go (hunk 15 above)
Comment (stgraber) [reply lxd/instances_put.go:80]: IIRC it's the time until all instances and tasks are running again.
During this time some API endpoints are responsive, others wait.

Comment (tomponline) [reply lxd/instances_put.go:80]: @stgraber ah yes so it is. Wonder if it would be better to reject API requests if LXD server isn't ready rather than blocking here (presumably this would apply to the other API route we use this channel too).
